homeassistant/components/renault/binary_sensor.py

```diff
@@ -1,58 +1,89 @@
 """Support for Renault binary sensors."""
 from __future__ import annotations
 
+from dataclasses import dataclass
+from typing import Generic, Optional, TypeVar
+
 from renault_api.kamereon.enums import ChargeState, PlugState
+from renault_api.kamereon.models import KamereonVehicleBatteryStatusData
 
 from homeassistant.components.binary_sensor import (
     DEVICE_CLASS_BATTERY_CHARGING,
     DEVICE_CLASS_PLUG,
     BinarySensorEntity,
+    BinarySensorEntityDescription,
 )
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.core import HomeAssistant
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
 
 from .const import DOMAIN
-from .renault_entities import RenaultBatteryDataEntity, RenaultDataEntity
+from .renault_entities import RenaultDataEntity, RenaultEntityDescription
 from .renault_hub import RenaultHub
 
 
+@dataclass
+class RenaultBinarySensorEntityDescription(
+    BinarySensorEntityDescription, RenaultEntityDescription
+):
+    """Class describing Renault binary sensor entities."""
+
+    on_value: StateType = None
+
+
+BINARY_SENSOR_TYPES: tuple[RenaultBinarySensorEntityDescription, ...] = (
+    RenaultBinarySensorEntityDescription(
+        key="plugged_in",
+        coordinator="battery",
+        data_key="plugStatus",
+        device_class=DEVICE_CLASS_PLUG,
+        entity_class="RenaultBatteryBinarySensor",
+        name="Plugged In",
+        on_value=PlugState.PLUGGED.value,
+    ),
+    RenaultBinarySensorEntityDescription(
+        key="charging",
+        coordinator="battery",
+        data_key="chargingStatus",
+        device_class=DEVICE_CLASS_BATTERY_CHARGING,
+        entity_class="RenaultBatteryBinarySensor",
+        name="Charging",
+        on_value=ChargeState.CHARGE_IN_PROGRESS.value,
+    ),
+)
+
+T = TypeVar("T")
+
+
 async def async_setup_entry(
     hass: HomeAssistant,
     config_entry: ConfigEntry,
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up the Renault entities from config entry."""
     proxy: RenaultHub = hass.data[DOMAIN][config_entry.entry_id]
-    entities: list[RenaultDataEntity] = []
+    entities: list[BinarySensorEntity] = []
     for vehicle in proxy.vehicles.values():
-        if "battery" in vehicle.coordinators:
-            entities.append(RenaultPluggedInSensor(vehicle, "Plugged In"))
-            entities.append(RenaultChargingSensor(vehicle, "Charging"))
+        for description in BINARY_SENSOR_TYPES:
+            if description.coordinator in vehicle.coordinators:
+                entity_class = globals()[description.entity_class]
```

Comment:
I don't really like this. Instead, if you want to make the class dynamic, move the descriptions to the end of the file and store `type[RenaultBinarySensor]` in a description field. Then you can just do `description.entity_class(...)`

Reply:
I am trying to implement this, and I feel I'm very close, but it won't initialise my entities and instead raises this error:
`__new__() missing 1 required positional argument: 'namespace'`

I have found this discussion but I'm not sure how I would implement it: https://stackoverflow.com/questions/37457207/python-typeerror-new-missing-1-required-positional-argument-namespace

Reply:
> I am trying to implement this, and I feel I'm very close, but it won't initialise my entities and instead raises this error:
> `__new__() missing 1 required positional argument: 'namespace'`

Can you push the changes so I can take a look? Maybe to a temporary branch?

Don't know if it would help, I did something similar in #55035. (Although there I had to import the descriptions from a separate file. I won't recommend that here.)

Reply:
I found the issue: I was using `description.entity_class = type(RenaultChargingSensor)` instead of `description.entity_class = RenaultChargingSensor` inside the descriptions.
It's all working now.